text/0000-negative-bounds.md

```diff
@@ -0,0 +1,682 @@
+
+// error. this impl conflict with T: Monad.
+impl MyFunctor for Vec { ... }
+```
+
+If we consider the aspect of HKTs as type constructors, they should behave similar to associated types — the output of the HKT should be considered opaque. This is out of scope of this RFC though.
+
+## No implicit specialization
+
+This RFC will not introduce implicit specialization. The following should still be rejected:
+
+```rust
+trait MyEq {}
+impl<T: PartialEq> MyEq for T {}
+impl<T: Eq> MyEq for T {}
+// error: conflicting implementations for trait `MyEq`
+```
+
+Explicit specialization is possible by excluding the specialized instances from the generic case.
+
+Inter-crate specialization is not possible with this method.
+
+Implicit specialization may be added back in the future, see the [Alternatives](#alternatives) section.
+
+## Miscellaneous
+
+* We should allow equality type bound syntax be also written as `where T == a`, to preserve the symmetry with `where T != a`.
+
+* The Clone trait may be changed to automatically implemented to those implementing Copy, as described in issue [#17884][17884]. This would require a `!Copy` bound to most custom implementations though. Similar for PartialEq/Eq and PartialOrd/Ord.
+
+    ```rust
+    impl<T: Copy> Clone for T {
+        fn clone(&self) -> T { *self }
+    }
+    impl<T, U> Clone for (T, U) where T: Clone, U: Clone, (T, U): !Copy {
+        fn clone(&self) -> (T, U) { (self.0.clone(), self.1.clone()) }
+    }
+    impl<T: Clone + !Copy> Clone for Option<T> {
+        fn clone(&self) -> Option<T> {
+            match *self {
+                None => None,
+                Some(ref e) => Some(e.clone()),
+            }
+        }
+    }
+    ```
+
+# Drawbacks
+
+* While negative bounds is a simple concept, it is also unintuitive. It is hard to understand `'a: !'b` without drawing figures for instance.
+
+* Negative bounds are purely a device to make the compiler happy about conflicting impl. It does not add any capability to the type it is bound on. To a human it may look like line noise.
+
+* Unlike specialization, it seems no other languages support negative bounds.
+
+* Extra syntax needs to be introduced.
+
+* Resolving an obligation becomes more complex.
+
+# Alternatives
+
+## Syntax
+
+### Negative bounds
+
+Instead `!` we may denote a negative bound with `-`. This is more consistent with `+` for combining bounds, and it's natural to introduce the shorthand `A + -B == A - B`:
+
+```rust
+impl<T: -Int> Trait for T {}
+impl<T: Int - Float> Trait for T {}
+```
+
+The advantage of `!` over `-` is that negative impls are already using `!`, and `!` seems more intuitive given that most old discussions raising this idea uses `!`.
+
+```rust
+impl<T> !Send for *const T {}
+impl<T: !Send> NotSync for T {}
+```
+
+### Inequality bounds
+
+Instead of `where T != u8`, we may write `where T: !u8` for an inequality bound. The advantage is we could add inequality to multiple types much more concisely:
```

Comment:
This wouldn’t really work because trait objects are types, so `T: !Reader` would be ambiguous between specifying that `T` does not implement `Reader` and that `T` is not an unsized `Reader` trait object type.